src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CustomMessageHandler;
+using Microsoft.CodeAnalysis.Remote.CustomMessageHandler;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed partial class RemoteCustomMessageHandlerService : BrokeredServiceBase, IRemoteCustomMessageHandlerService
+{
+#if !NETSTANDARD2_0
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly ConcurrentDictionary<string, CustomMessageHandlerExtension> _extensions = new();
+
+    private readonly System.Runtime.Loader.AssemblyLoadContext _defaultLoadContext
+        = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(typeof(RemoteCustomMessageHandlerService).Assembly)
+        ?? throw new InvalidOperationException($"Cannot get assembly load context for {nameof(RemoteCustomMessageHandlerService)}.");
+#endif
+
+    internal sealed class Factory : FactoryBase<IRemoteCustomMessageHandlerService>
+    {
+        protected override IRemoteCustomMessageHandlerService CreateService(in ServiceConstructionArguments arguments)
+            => new RemoteCustomMessageHandlerService(arguments);
+    }
+
+    public RemoteCustomMessageHandlerService(in ServiceConstructionArguments arguments)
+        : base(arguments)
+    {
+    }
+
+    public ValueTask<HandleCustomMessageResponse> HandleCustomMessageAsync(
+        Checksum solutionChecksum,
+        string assemblyPath,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        ImmutableArray<LinePosition> positions,
+        CancellationToken cancellationToken)
+    {
+#if NETSTANDARD2_0
+        throw new InvalidOperationException("Custom handlers are only supported in the servicehub host");
+#else
+
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        Requires.NotNullOrEmpty(assemblyPath);
+        Requires.NotNullOrEmpty(typeFullName);
+        Requires.NotNullOrEmpty(jsonMessage);
+        Requires.NotNullOrEmpty(positions);
+
+        return RunServiceAsync(solutionChecksum, async solution =>
+        {
+            Document? document = null;
+            if (documentId is not null)
+            {
+                document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(document);
+            }
+
+            try
+            {
+                var extension = _extensions.GetOrAdd(assemblyPath, _ =>
```

Comment:
I don't think there is an issue here, but I'm not intimately familiar with all the details of ALC creation - it is possible that this lambda runs multiple times if there are concurrent calls for the same assembly path.

I think that is fine as the unused ALC would just get discarded?  However it might be a bit simpler to grok if used a lock or similar instead.